teps/0056-pipelines-in-pipelines.md

```diff
@@ -0,0 +1,358 @@
+above workflow is simple, it has limitations in terms of composability, reusability and failure strategies in `Pipelines`. 
+
+#### Reusability and Composability
+
+Today, a subset of `Tasks` in a `Pipeline` cannot be grouped together and distributed as a unit of execution within the
+`Pipeline`. As such, users have to specify those related `Tasks` separately and repeatedly across many `Pipelines`.
+
+Users need to define and share a set of `Tasks` as a complete unit of execution.
+
+#### Failure Strategies
+
+Today, when a `Task` in a `Branch` fails, it stops the execution of unrelated `Branches` and the `Pipeline` as a whole.
+When `Tasks` are specified in independent `Branches`, there are no dependencies between them, so users may expect 
+that a failure in one `Branch` would not stop the execution of the other `Branch`.
+
+Users need a way to prevent the failure of unrelated `Tasks` from affecting the execution of set of related `Tasks` that
+should execute to completion once the first one has started. 
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Defining a set of `Tasks` as a complete unit of execution
+- Decoupling failures in unrelated sets of `Tasks`
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Ignoring `Task` failures, which is addressed in [TEP-0050 Ignore Task Failures](https://github.com/tektoncd/community/blob/main/teps/0050-ignore-task-failures.md)
+- Composing `Tasks` in Tasks`, which is addressed in [TEP-0044 Composing Tasks with Tasks](https://github.com/tektoncd/community/pull/316)
+
+### Use Cases
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+* As a `Pipeline` author, I need to define a set of `Tasks` as a complete unit of execution that I can share across 
+  `Pipelines`.
+  
+For example, I can define and distribute `linting` set of `Tasks` and `testing` set of `Tasks` where:
+
+`linting` is made up of `lint` and `report-linter-output` tasks:
+``` 
+          lint                     
+           |                           
+           v
+   report-linter-output 
+```
+
+`testing` is made up of `unit-tests`, `integration-tests` and `report-test-results` `Tasks`:
+```                           
+       unit-tests
+           |                           
+           v                           
+   integration-tests 
+           |
+           v 
+  report-test-results
+```
+
+* As a `Pipeline` author, I need to decouple failures in unrelated sets of `Tasks`.
```

Comment:
Decoupling of such failures is only possible because `linting` sub pipeline and `tests` sub pipeline start executing at the same time. It wouldn't be possible to decouple if `tests` sub pipeline is scheduled after `linting` as an example. I think the biggest motivation and use case for this feature would be to construct the sharable pipelines. I would make that (shareability) as the main use case in addition to this.

Reply:
thanks for the review @pritidesai! 

> Decoupling of such failures is only possible because `linting` sub pipeline and `tests` sub pipeline start executing at the same time. It wouldn't be possible to decouple if `tests` sub pipeline is scheduled after `linting` as an example

if the `testing` subpipeline is scheduled after the `linting` subpipeline, that means there are dependencies between them (they're in the same branch), so the failure in `linting` can impact the `testing` subpipeline

what I was trying to address there is when the subpipelines are executed in parallel, when `linting` subpipeline fails then `testing` subpipeline would currently stop execution but we'd want it to run to completion so we get the test results (because there are no dependencies with the `linting` subpipeline`)

> I think the biggest motivation and use case for this feature would be to construct the sharable pipelines. I would make that (shareability) as the main use case in addition to this.

agreed, i tried to capture that in `reusability`...maybe I need to expound on it further or is `shareability` separate from that? please let me know

Reply:
> if the `testing` subpipeline is scheduled after the `linting` subpipeline, that means there are dependencies between them (they're in the same branch), so the failure in `linting` can impact the `testing` subpipeline

not necessarily, lets take a look at a bit more complex graph such as: 

```
short-running-sub-pipeline(linting)        long-running-sub-pipeline(build)
                                                         |
                                              sub-pipeline(test) 

```

Here, if the `short-running-sub-pipeline` has a failure, the `sub-pipeline` which has no direct connection to it will not be executed.

Reply:
> agreed, i tried to capture that in reusability...maybe I need to expound on it further or is shareability separate from that? please let me know

Both are aligned, no need to expound on it further.

Reply:
> > if the `testing` subpipeline is scheduled after the `linting` subpipeline, that means there are dependencies between them (they're in the same branch), so the failure in `linting` can impact the `testing` subpipeline
> 
> not necessarily, lets take a look at a bit more complex graph such as:
> 
> ```
> short-running-sub-pipeline(linting)        long-running-sub-pipeline(build)
>                                                          |
>                                               sub-pipeline(test) 
> ```
> 
> Here, if the `short-running-sub-pipeline` has a failure, the `sub-pipeline` which has no direct connection to it will not be executed.

yes that's right, so a user who wants `sub-pipeline(tests)` to execute when `short-running-sub-pipeline(linting)` fails would combine `build-and-test-sub-pipeline` made up of `long-running-sub-pipeline(build)` and `sub-pipeline(test)`

I could clarify that to "when there's a failure in unrelated Tasks, continue the execution of the set of Tasks once the first one has started" -- so in the above scenario, `long-running-sub-pipeline(build)` would run to completion -- what do you think?

Reply:
thanks for the updates @jerop, looks great 🙏